File: Python/Composite/Composite.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
'''
Composite Pattern
Author: Kei Nakata
Data: Oct.10.2014
'''
import abc
import exceptions

class Component(object):
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def __init__(self, name):
        pass

    @abc.abstractmethod
    def add(self, child):
        pass

    @abc.abstractmethod
    def remove(self, index):
        pass

    @abc.abstractmethod
    def getChild(self, index):
        pass

    @abc.abstractmethod
    def show(self):
        pass

class Composite(Component):

    def __init__(self, name):
        self.children = []

    def add(self, child):
        self.children.append(child)

    def remove(self, index):
        del self.children[index]

    def getChild(self, index):
        return self.children[index]

    def show(self):
        for child in self.children:
            child.show()

class Leaf(Component):
    count = 0
    def __init__(self, name):
        self.name = name
        Leaf.count = Leaf.count + 1
        self.number = Leaf.count
    def add(self):
        raise exceptions.RuntimeError("can not add item to leaf")

    def remove(self):
        raise exceptions.RuntimeError("can not remove item through leaf class")

    def getChild(self):
        raise exceptions.RuntimeError("leaf does not have child")

    def show(self):
        print self.number, self.name

if __name__ == '__main__':
    container = Composite('box')
    small_container = Composite('small box')
    small_container.add(Leaf('chicken'))
    small_container.add(Leaf('beaf'))
    small_container.add(Leaf('pork'))
    container.add(Leaf('apple'))
    container.add(Leaf('orange'))
    container.add(Leaf('pear'))
    container.add(small_container)
    container.show()
    print
    container.remove(1)
    container.show()

File: C++/Composite/Composite.cpp
/*
 * Composite Pattern
 * Author: Kei Nakata
 * Data: Oct.6.2014
 * Compose objects into tree structures to represent part-whole
 * hierarchies. Composite lets clients treat individual objects
 * and composition of objects unifomly.
 */
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <map>
using std::cout;
using std::endl;
using std::string;
using std::map;
using std::pair;

class Component
{
protected:
  string _name;
public:
  Component(const string& name) : _name(name) {}
  virtual ~Component() {}
  virtual void show() = 0;
};

class Composite :public Component
{
  map<int, Component*> children;
public:
  Composite(const string& name) : Component(name) {}
  ~Composite()
  {
    map<int, Component*>::iterator it;
    for(it = children.begin(); it != children.end(); ++it)
      {
        delete it->second;
      }
  };

  void add(Component* child)
  {
    static int index = 0;
    children.insert(pair<int, Component*>(index, child));
    ++index;
  }

  void show()
  {
    map<int, Component*>::iterator it = children.begin();
    for(it = children.begin(); it != children.end(); ++it)
      {
        it->second->show();
      }
  }

};

class Leaf:public Component
{
public:
  Leaf(const string& name) : Component(name) {}
  void show()
  {
    cout << this->_name << endl;
  }
};



int main()
{
  Composite* container= new Composite("box");
  Composite* small_container = new Composite("small_box");
  container->add(new Leaf("apple"));
  container->add(new Leaf("pear"));
  container->add(new Leaf("orange"));
  small_container->add(new Leaf("beaf"));
  small_container->add(new Leaf("chicken"));
  small_container->add(new Leaf("pork"));
  container->add(small_container);
  container->show();
  delete container;

  return EXIT_SUCCESS;
}
